### PR TITLE
Fix status response in write client

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Fix status response in write client
This PR is cut from https://github.com/project-chip/connectedhomeip/pull/19356

#### Change overview
Then onMessageReceived in WriteClient would send status response message with invalid action when there is error happening in incoming message.

#### Testing
WriteClient sends write requests, and Injects malformed write response, status response, unknown message in client side, and client would send status response message with invalid action.

### CODE REVIEW

File: src/app/tests/TestWriteInteraction.cpp [hunk 1]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
     engine->Shutdown();
 }
 
+// Write Client sends the write request, and process the unknown message error via OnMessageReceived to close the client.
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:555]:
```suggestion
// Write Client sends a write request, receives an unexpected message type, sends a status response to that.
```

File: src/app/tests/TestWriteInteraction.cpp [hunk 2]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
...
+
+    app::WriteClient writeClient(engine->GetExchangeManager(), &callback, Optional<uint16_t>::Missing());
+
+    System::PacketBufferHandle buf = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    AddAttributeDataIB(apSuite, apContext, writeClient);
+
+    NL_TEST_ASSERT(apSuite, callback.mOnSuccessCalled == 0 && callback.mOnErrorCalled == 0 && callback.mOnDoneCalled == 0);
+
+    ctx.GetLoopback().mSentMessageCount                 = 0;
+    ctx.GetLoopback().mNumMessagesToDrop                = 3;
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:579]: Why 3, if we only drop 1 in practice?

File: src/app/tests/TestWriteInteraction.cpp [hunk 3]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
...
+    ctx.GetLoopback().mNumMessagesToDrop                = 0;
+    ctx.GetLoopback().mNumMessagesToAllowBeforeDropping = 0;
+    ctx.GetLoopback().mDroppedMessageCount              = 0;
+    err = writeClient.OnMessageReceived(writeClient.mExchangeCtx.Get(), payloadHeader, std::move(msgBuf));
+    NL_TEST_ASSERT(apSuite, err == CHIP_ERROR_INVALID_MESSAGE_TYPE);
+    ctx.DrainAndServiceIO();
+    NL_TEST_ASSERT(apSuite, callback.mError == CHIP_ERROR_INVALID_MESSAGE_TYPE);
+    NL_TEST_ASSERT(apSuite, callback.mOnSuccessCalled == 0 && callback.mOnErrorCalled == 1 && callback.mOnDoneCalled == 1);
+
+    // Client sents status report with invalid action, server's exchange has been closed, and it would send MRP Ack
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:611]:
```suggestion
    // Client sents status report with invalid action, server's exchange has been closed, so all it sends is an MRP Ack
```

File: src/app/tests/TestWriteInteraction.cpp [hunk 4]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
...
+
+    engine->Shutdown();
+    ctx.ExpireSessionAliceToBob();
+    ctx.ExpireSessionBobToAlice();
+    ctx.CreateSessionAliceToBob();
+    ctx.CreateSessionBobToAlice();
+}
+
+// Write Client sends the write request, and process the invalid write response message error via OnMessageReceived to close the
+// client.
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:622]:
```suggestion
// Write Client sends a write request, receives a malformed write response message, sends a Status Report.
```

File: src/app/tests/TestWriteInteraction.cpp [hunk 5]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
...
+
+    app::WriteClient writeClient(engine->GetExchangeManager(), &callback, Optional<uint16_t>::Missing());
+
+    System::PacketBufferHandle buf = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    AddAttributeDataIB(apSuite, apContext, writeClient);
+
+    NL_TEST_ASSERT(apSuite, callback.mOnSuccessCalled == 0 && callback.mOnErrorCalled == 0 && callback.mOnDoneCalled == 0);
+
+    ctx.GetLoopback().mSentMessageCount                 = 0;
+    ctx.GetLoopback().mNumMessagesToDrop                = 3;
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:646]: Again, why 3?

File: src/app/tests/TestWriteInteraction.cpp [hunk 6]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
...
+    ctx.GetLoopback().mNumMessagesToDrop                = 0;
+    ctx.GetLoopback().mNumMessagesToAllowBeforeDropping = 0;
+    ctx.GetLoopback().mDroppedMessageCount              = 0;
+    err = writeClient.OnMessageReceived(writeClient.mExchangeCtx.Get(), payloadHeader, std::move(msgBuf));
+    NL_TEST_ASSERT(apSuite, err == CHIP_ERROR_END_OF_TLV);
+    ctx.DrainAndServiceIO();
+    NL_TEST_ASSERT(apSuite, callback.mError == CHIP_ERROR_END_OF_TLV);
+    NL_TEST_ASSERT(apSuite, callback.mOnSuccessCalled == 0 && callback.mOnErrorCalled == 1 && callback.mOnDoneCalled == 1);
+
+    // Client sents status report with invalid action, server's exchange has been closed, and it would send MRP Ack
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:678]:
```suggestion
    // Client sents status report with invalid action, server's exchange has been closed, so all it does is send an MRP ack.
```

Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:678]: This needs a TODO comment about checking that the server gets the right status....

File: src/app/tests/TestWriteInteraction.cpp (hunk 3 above)
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:611]: This needs a TODO comment about checking that the server gets the right status....

File: src/app/tests/TestWriteInteraction.cpp [hunk 7]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
...
+
+    engine->Shutdown();
+    ctx.ExpireSessionAliceToBob();
+    ctx.ExpireSessionBobToAlice();
+    ctx.CreateSessionAliceToBob();
+    ctx.CreateSessionBobToAlice();
+}
+
+// Write Client sends the write request, and process the malformed status response message error via OnMessageReceived to close the
+// client.
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:689]:
```suggestion
// Write Client sends a write request, receives a malformed status response message.
```

File: src/app/tests/TestWriteInteraction.cpp [hunk 8]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
...
+
+    app::WriteClient writeClient(engine->GetExchangeManager(), &callback, Optional<uint16_t>::Missing());
+
+    System::PacketBufferHandle buf = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    AddAttributeDataIB(apSuite, apContext, writeClient);
+
+    NL_TEST_ASSERT(apSuite, callback.mOnSuccessCalled == 0 && callback.mOnErrorCalled == 0 && callback.mOnDoneCalled == 0);
+
+    ctx.GetLoopback().mSentMessageCount                 = 0;
+    ctx.GetLoopback().mNumMessagesToDrop                = 3;
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:713]: Why 3?

File: src/app/tests/TestWriteInteraction.cpp [hunk 9]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
...
+    ctx.GetLoopback().mNumMessagesToDrop                = 0;
+    ctx.GetLoopback().mNumMessagesToAllowBeforeDropping = 0;
+    ctx.GetLoopback().mDroppedMessageCount              = 0;
+    err = writeClient.OnMessageReceived(writeClient.mExchangeCtx.Get(), payloadHeader, std::move(msgBuf));
+    NL_TEST_ASSERT(apSuite, err == CHIP_ERROR_END_OF_TLV);
+    ctx.DrainAndServiceIO();
+    NL_TEST_ASSERT(apSuite, callback.mError == CHIP_ERROR_END_OF_TLV);
+    NL_TEST_ASSERT(apSuite, callback.mOnSuccessCalled == 0 && callback.mOnErrorCalled == 1 && callback.mOnDoneCalled == 1);
+
+    // Client sents status report with invalid action, server's exchange has been closed, and it would send MRP Ack
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:745]: This needs a TODO comment about checking that the server gets the right status....

Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:745]:
```suggestion
    // Client sents status report with invalid action, server's exchange has been closed, so all it sends is an MRP ack.
```

File: src/app/tests/TestWriteInteraction.cpp [hunk 10]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
...
+
+    engine->Shutdown();
+    ctx.ExpireSessionAliceToBob();
+    ctx.ExpireSessionBobToAlice();
+    ctx.CreateSessionAliceToBob();
+    ctx.CreateSessionBobToAlice();
+}
+
+// Write Client sends the write request, and process the busy status response message error via OnMessageReceived to close the
+// client.
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:756]:
```suggestion
// Write Client sends a write request, receives a busy status response message.
```

File: src/app/tests/TestWriteInteraction.cpp [hunk 11]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
...
+
+    app::WriteClient writeClient(engine->GetExchangeManager(), &callback, Optional<uint16_t>::Missing());
+
+    System::PacketBufferHandle buf = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    AddAttributeDataIB(apSuite, apContext, writeClient);
+
+    NL_TEST_ASSERT(apSuite, callback.mOnSuccessCalled == 0 && callback.mOnErrorCalled == 0 && callback.mOnDoneCalled == 0);
+
+    ctx.GetLoopback().mSentMessageCount                 = 0;
+    ctx.GetLoopback().mNumMessagesToDrop                = 3;
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:780]: Why 3?

File: src/app/tests/TestWriteInteraction.cpp [hunk 12]
```diff
@@ -543,6 +552,274 @@ void TestWriteInteraction::TestWriteRoundtrip(nlTestSuite * apSuite, void * apCo
...
+    ctx.GetLoopback().mNumMessagesToDrop                = 0;
+    ctx.GetLoopback().mNumMessagesToAllowBeforeDropping = 0;
+    ctx.GetLoopback().mDroppedMessageCount              = 0;
+    err = writeClient.OnMessageReceived(writeClient.mExchangeCtx.Get(), payloadHeader, std::move(msgBuf));
+    NL_TEST_ASSERT(apSuite, err == CHIP_IM_GLOBAL_STATUS(Busy));
+    ctx.DrainAndServiceIO();
+    NL_TEST_ASSERT(apSuite, callback.mError == CHIP_IM_GLOBAL_STATUS(Busy));
+    NL_TEST_ASSERT(apSuite, callback.mOnSuccessCalled == 0 && callback.mOnErrorCalled == 1 && callback.mOnDoneCalled == 1);
+
+    // Client sents status report with invalid action, server's exchange has been closed, and it would send MRP Ack
```
Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:813]: This needs a TODO comment about checking that the server gets the right status....

Comment (bzbarsky-apple) [inline src/app/tests/TestWriteInteraction.cpp:813]:
```suggestion
    // Client sents status report with invalid action, server's exchange has been closed, so it just sends an MRP ack.
```

File: src/app/tests/TestWriteInteraction.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:579]: yes, it should be 1. typo here. updated.

File: src/app/tests/TestWriteInteraction.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:555]: done, thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 4 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:622]: done, thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 5 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:646]: should be 1. updated. thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 9 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:745]: updated, thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 11 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:780]: updated with 1, thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 10 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:756]: done, thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 12 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:813]: done, thanks

Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:813]: added, thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:611]: Done, thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 6 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:678]: done, thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 8 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:713]: update with 1, thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 9 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:745]: Done, thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:611]: done, thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 7 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:689]: done, thanks

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

**Pull request merged**